### PR TITLE
F: Add benevolence import support

### PR DESCRIPTION
**Pull request by guymaness2**
### Description 

Added support for Benevolence imports via CSV and Binary components of Bulldozer

##### What does the change add or fix?

* Added support for Benevolence Request CSV importing.
* Added support for Benevolence Result CSV importing.
* Added support for Benevolence Document Binary importing.  

**New Settings:**

none

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

CSV files can now be used to import BenevolenceRequests and BenevolenceResults (see respective sample CSV's in the sample CSV section). Benevolence documents can also be imported through binary mode. The name of the imported zip file must begin with "BenevolenceRequestDocuments" for the binary import tool to recognize it. File names should follow the structure of REQUESTFOREIGNID_DESIREDFILENAME_DOCUMENTFOREIGNID.EXTENTION (ex. 2932_electricbill_4321.jpg would result in a BenevolenceDocument being created with the foreignId of 4321 and filename of "electricbill.jpg" connected to the previously imported BenevolenceRequest with ForeignId of 2932).  I'll update the wiki when I get a chance to give better descriptions of how to configure the csv files and document file names.

---------

##### Who reported, requested, or paid for the change?

Victory World Church

---------

##### Does this update or add options to the block UI?

none

---------

##### What files does it affect?

* Bulldozer.BinaryFile/Bulldozer.BinaryFile.csproj
* Bulldozer.BinaryFile/Maps/BenevolenceRequestDocuments.cs
* Bulldozer.CSV/Bulldozer.CSV.csproj
* Bulldozer.CSV/CSVComponent.cs
* Bulldozer.CSV/CSVInstance.cs
* Bulldozer.CSV/Maps/Benevolence.cs
* Bulldozer.CSV/SampleCSVs/06 - Contribution.csv
* Bulldozer.CSV/SampleCSVs/26 - BenevolenceRequest.csv
* Bulldozer/Utility/AddMethods.cs

---------

##### Is it a breaking change for other versions/clients?

none

### CODE REVIEW

File: Bulldozer.CSV/Maps/Benevolence.cs [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+                var benevolenceRequestCountry = row[BenevolenceRequestCountry];
+
+                //
+                // Verify we have the minimum required information for a valid BenevolenceRequest in the csv file.
+                //
+                if ( string.IsNullOrWhiteSpace( benevolenceRequestText )
+                    || ( string.IsNullOrWhiteSpace( benevolenceRequestRequestedById ) && ( string.IsNullOrWhiteSpace( benevolenceRequestFirstName ) || string.IsNullOrWhiteSpace( benevolenceRequestLastName ) ) ) )
+                {
+
+                    throw new System.Collections.Generic.KeyNotFoundException( $"Benevolence Request {benevolenceRequestId} is missing information. BenevolenceRequestText and either BenevolenceRequestRequestedById or both BenevolenceRequestFirstName and BenevolenceRequestLastName are required. ", null );
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Benevolence.cs:94]: Does this kill the entire import after the loop? Can this scenario be tested?

File: Bulldozer.CSV/Maps/Benevolence.cs [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+                var exists = false;
+                if ( importedBenevolenceRequests > 0 )
+                {
+                    exists = benevolenceRequestService.Queryable().AsNoTracking().Any( r => r.ForeignKey == benevolenceRequestId );
+                }
+
+                if ( !exists )
+                {
+                    var email = string.Empty;
+                    var firstName = benevolenceRequestFirstName;
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Benevolence.cs:109]: Are these used anywhere?

Comment (nateh777) [inline Bulldozer.CSV/Maps/Benevolence.cs:109]:
```suggestion
```

File: Bulldozer.CSV/Maps/Benevolence.cs [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+                    var lastName = benevolenceRequestLastName;
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Benevolence.cs:110]:
```suggestion
```

File: Bulldozer.CSV/Maps/Benevolence.cs [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+                BenevolenceRequest benevolenceRequest = null;
+                if ( benevolenceRequestService.Queryable().AsNoTracking().Any( r => r.ForeignKey == benevolenceResultRequestId ) )
+                {
+                    benevolenceRequest = benevolenceRequestService.Queryable().AsNoTracking().FirstOrDefault( r => r.ForeignKey == benevolenceResultRequestId );
+                }
+
+                //
+                // Verify the Benevolence Request exists.
+                //
+                if ( benevolenceRequest.Id < 1 )
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Benevolence.cs:329]:
```suggestion
                if ( benevolenceRequest == null || benevolenceRequest.Id < 1 )
```

File: Bulldozer.CSV/SampleCSVs/26 - BenevolenceRequest.csv [hunk 5]
```diff
@@ -0,0 +1,5 @@
+﻿BenevolenceRequestText,BenevolenceRequestDate,BenevolenceRequestId,BenevolenceRequestFirstName,BenevolenceRequestLastName,BenevolenceRequestEmail,BenevolenceRequestCreatedById,BenevolenceRequestCreatedDate,BenevolenceRequestRequestedById,BenevolenceRequestCaseWorkerId,BenevolenceRequestCellPhone,BenevolenceRequestHomePhone,BenevolenceRequestWorkPhone,BenevolenceRequestGovernmentId,BenevolenceRequestProvidedNextSteps,BenevolenceRequestStatus,BenevolenceRequestResultSummary,BenevolenceRequestAddress,BenevolenceRequestAddress2,BenevolenceRequestCity,BenevolenceRequestState,BenevolenceRequestZip,BenevolenceRequestCountry
+Requesting help paying rent,2020-04-09 00:00:00.000,2,Barney,Rubble,barneyr@bedrock.com,10477,2020-04-09 15:46:30.210,10444,NULL,,,,,,Approved,,NULL,NULL,NULL,NULL,NULL,NULL
```
Comment (nateh777) [inline Bulldozer.CSV/SampleCSVs/26 - BenevolenceRequest.csv:2]: Can we test without null's and sanitize this csv without null?

Comment (nateh777) [reply Bulldozer.CSV/SampleCSVs/26 - BenevolenceRequest.csv:2]: Also we need a BenevolenceResult.csv sample.

File: Bulldozer.CSV/Maps/Benevolence.cs (hunk 1 above)
Comment (guymaness2) [reply Bulldozer.CSV/Maps/Benevolence.cs]: It indeed did kill the process. Just committed code to change all exception handling for the benevolence map to provide a report to the console as well as enter an exception into the exception log and keep moving.

File: Bulldozer.CSV/SampleCSVs/26 - BenevolenceRequest.csv (hunk 5 above)
Comment (guymaness2) [reply Bulldozer.CSV/SampleCSVs/26 - BenevolenceRequest.csv]: Done and done.

**Review by nateh777 — APPROVED** [review]

**Pull request merged**